### PR TITLE
vtk-cdat recipe

### PR DESCRIPTION
**Pull request by doutriaux1**

### CODE REVIEW

**Comment by doutriaux1** [comment]
@danlipsa now we reach a timeout... How can we speed up things?

**Comment by doutriaux1** [comment]
@sankhesh same here. The build finishes but the conda `rpath` bit for mac takes forever (same on my local mac) and we run into a time out. I'm at `-j8` so i don't think we can speed things up much here. @dnadeau4 ideas welcome as well, you know conda-forge pretty well. @ocefpaf you helped us before on some other PR. Any idea on how to bypass travis timeout?

File: recipes/vtk-cdat/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+source:
+    git_rev: 1c03d1d03b1c12d99dd090ba4c5ca078e2d9612f
+    git_url: git://github.com/Kitware/VTK.git
+
+build:
+    skip: True  # [win or py3k]
+    number: 0
+    detect_binary_files_with_prefix: true
+    has_prefix_files:  # [win]
+        - Lib/site-packages/vtk/__init__.py  # [win]
```
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: Are these really necessary?

File: recipes/vtk-cdat/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,59 @@
+{% set version = "2.10" %}
+{% set name = "vtk-cdat" %}
+
+package:
+    name: {{ name }}
+    version: 7.1.0.{{ version }}
+
+source:
+    git_rev: 1c03d1d03b1c12d99dd090ba4c5ca078e2d9612f
+    git_url: git://github.com/Kitware/VTK.git
```
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: Use an official tarball instead of a git archive.

File: recipes/vtk-cdat/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,59 @@
+{% set version = "2.10" %}
+{% set name = "vtk-cdat" %}
+
+package:
+    name: {{ name }}
+    version: 7.1.0.{{ version }}
```
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: Change `version` jinja variable to `7.1.0.2.10`

File: recipes/vtk-cdat/build.sh [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+fi
+
+COMMAND="cmake .. \
+    ${COMMON_ARGS} \
+    ${VTK_ARGS} \
+    ${MPI_ARGS}"
+echo $COMMAND
+eval ${COMMAND}
+
+make -j8
```
Comment (isuruf) [inline recipes/vtk-cdat/build.sh]: `make -j${CPU_COUNT}`

File: recipes/vtk-cdat/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+        - libgcc  # [linux]
+        - toolchain
+    run:
+        - python
+        - proj4 >4.9.2
+        - hdf5 1.8.18
+        - ffmpeg
+        - freetype
+        - x264 <20151006.2245
+        - gcc  # [linux]
```
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: Is `gcc` needed at runtime?

File: .travis.yml [hunk 6]
```diff
@@ -2,7 +2,7 @@
 language: generic
 
 os: osx
-osx_image: xcode6.4
+osx_image: xcode8.2
```
Comment (isuruf) [inline .travis.yml]: Why this change?

File: recipes/vtk-cdat/build.sh [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+        -DVTK_INSTALL_PYTHON_MODULE_DIR=${SP_DIR} \
+        -DModule_vtkRenderingMatplotlib=ON \
+        -DVTK_Group_Web:BOOL=ON \
+        -DLIBPROJ4_INCLUDE_DIR:PATH=${PREFIX}/include \
+        -DVTK_RENDERING_BACKEND=OpenGL \
+        -DVTK_USE_SYSTEM_ZLIB:BOOL=ON \
+        -DVTK_USE_SYSTEM_LIBXML2:BOOL=ON \
+        -DVTK_USE_SYSTEM_HDF5:BOOL=ON \
+        -DVTK_USE_SYSTEM_NETCDF:BOOL=ON \
+        -DVTK_USE_SYSTEM_FREETYPE:BOOL=OFF \
```
Comment (isuruf) [inline recipes/vtk-cdat/build.sh]: Shouldn't this be `ON`?

File: recipes/vtk-cdat/meta.yaml [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        - Lib/site-packages/vtk/__init__.py  # [win]
+
+requirements:
+    build:
+        - cmake
+        - python
+        - proj4 >4.9.2
+        - hdf5 1.8.18
+        - ffmpeg
+        - freetype
```
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: These need to be pinned

**Comment by isuruf** [comment]
@doutriaux1, one way would be to use travis-ci caching to reduce the build time. It's still not implemented yet though, https://github.com/conda-forge/conda-smithy/pull/526

**Comment by ocefpaf** [comment]
> @ocefpaf you helped us before on some other PR. Any idea on how to bypass travis timeout?

I did not look at the recipe yet (traveling today) but reducing the amount of tests and using multiple cores (-j $CPU_COUNY) sometimes help.

> one way would be to use travis-ci caching to reduce the build time

caching may help in some cases, but from Travis-CI own docs:

    Large files that are quick to install but slow to download do not benefit from caching,
    as they take as long to download from the cache as from the original source

TL;DR I only had speedups with Travis-CI caching when the conda env I created was really big and Travis-CI was faster than Anaconda.org to "download" the cache.

**Comment by doutriaux1** [comment]
@isuruf yes that would work, another way I was thinking off would be to edit your configs! And rather than running one script copy the script into `.travis.yml` that would give us more `jobs` as far as travis is concerned and it might help getting under the 50mn per job limit.
@ocefpaf thanks for the pointers unfortunately we're already using the `-j` and we do not get to test part... It finishes the actual build part but times out because of conda's `otool` step to fix all the `.so` files.

File: recipes/vtk-cdat/meta.yaml (hunk 3 above)
Comment (sankhesh) [inline recipes/vtk-cdat/meta.yaml]: @isuruf Out of curiosity, why?

File: recipes/vtk-cdat/meta.yaml (hunk 2 above)
Comment (sankhesh) [inline recipes/vtk-cdat/meta.yaml]: VTK doesn't put out nightly tarballs. This recipe enables us to use a VTK version that may be ahead of the official release.

File: recipes/vtk-cdat/meta.yaml (hunk 3 above)
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: These jinja variables are declared at the top so that when you want to update you just change the version variable. Here, if the version needs to be updated it has to be changed in several places which defeats the purpose.

File: recipes/vtk-cdat/meta.yaml (hunk 2 above)
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: It's better to package stable versions than git master. Is there a reason for needing git master?

**Comment by isuruf** [comment]
> @isuruf yes that would work, another way I was thinking off would be to edit your configs! And rather than running one script copy the script into .travis.yml that would give us more jobs as far as travis is concerned and it might help getting under the 50mn per job limit.

I'm not sure I understand.

> TL;DR I only had speedups with Travis-CI caching when the conda env I created was really big and Travis-CI was faster than Anaconda.org to "download" the cache.

@ocefpaf, I was thinking of using ccache to speedup the building. https://github.com/conda-forge/conda-smithy/pull/526

File: recipes/vtk-cdat/meta.yaml (hunk 2 above)
Comment (doutriaux1) [inline recipes/vtk-cdat/meta.yaml]: @sankhesh we can use our fork. @isuruf as Sankhesh explained we are using VTK and are usually adding filters or fixing VTK bug so our stable releases are based on a stable VTK SHA1 but not an official. Our nightlies (in between version) will not be hosted on conda-forge but on our conda channel (`uvcdat`). I will clean all this up as soon as we have a passing recipe, right now the focus is on getting it to pass travis.

File: recipes/vtk-cdat/meta.yaml (hunk 1 above)
Comment (doutriaux1) [inline recipes/vtk-cdat/meta.yaml]: copied from official VTK recipe, hoping to move to `win` eventually, but yes we could remove for now.

File: recipes/vtk-cdat/build.sh (hunk 7 above)
Comment (doutriaux1) [inline recipes/vtk-cdat/build.sh]: yes as well, might even save us time! Thanks!

File: recipes/vtk-cdat/build.sh (hunk 4 above)
Comment (doutriaux1) [inline recipes/vtk-cdat/build.sh]: too many commits from my local test to master. Thanks!

File: .travis.yml (hunk 6 above)
Comment (doutriaux1) [inline .travis.yml]: Look at the builds, XCode 6 and 7 fail with some COCOA/CARBON errors.

**Comment by doutriaux1** [comment]
@isuruf look at: https://github.com/conda-forge/staged-recipes/blob/master/.travis.yml#L34
this groups all the steps into one! therefore limiting travis build time to 50mn for the whole build. You might want this for conda-forge though to limit each build... But I think on travis there is not a time limit on a build itself, only 50mn hard limit on each steps/job of the build.

**Comment by isuruf** [comment]
@doutriaux1, you are only building python 2.7 for now right? How can this be divided into multiple travis jobs?

**Comment by ocefpaf** [comment]
> @ocefpaf, I was thinking of using ccache to speedup the building. conda-forge/conda-smithy#526

That might help but I am afraid that we do need a "stronger" Travis-CI account. (We are working on it...)

**Comment by doutriaux1** [comment]
@isuruf i think if I copy all the commands in .travis.yml then I get 50mn per command, not 50mn for the whole thing.

File: recipes/vtk-cdat/build.sh (hunk 4 above)
Comment (doutriaux1) [inline recipes/vtk-cdat/build.sh]: @isuruf also I read somewhere than travis/mac always return one for this.

Comment (isuruf) [inline recipes/vtk-cdat/build.sh]: No. we set it to 2 here, https://github.com/conda-forge/conda-forge-build-setup-feedstock/blob/master/recipe/run_conda_forge_build_setup_osx#L3

File: recipes/vtk-cdat/meta.yaml (hunk 1 above)
Comment (isuruf) [inline recipes/vtk-cdat/meta.yaml]: See this https://github.com/conda-forge/vtk-feedstock/blob/master/recipe/meta.yaml
